lib/dialogs/settings.jsx

```diff
@@ -37,6 +38,72 @@ export const SettingsDialog = React.createClass({
     });
   },
 
+  onSignOutRequested() {
+    // Safety first! Check for any unsynced notes before signing out.
+    const { onSignOut, noteBucket } = this.props;
+    const { notes } = this.props.appState;
+    const { getVersion } = noteBucket;
+
+    noteBucket.hasLocalChanges((error, hasChanges) => {
+      if (hasChanges) {
+        this.showUnsyncedWarning();
+        return;
+      }
+
+      // Also check persisted store for any notes with version 0
+      const noteHasSynced = note =>
+        new Promise((resolve, reject) =>
+          getVersion(note.id, (e, v) => (e || v === 0 ? reject() : resolve()))
+        );
+
+      Promise.race(notes.map(noteHasSynced)).then(
```

Comment:
Looks like I commented on an outdated change but the `Promise.race` remains.

It seems you want to check if any of the notes have a version of `0` and if any of them do have a version of `0` then you will want to show `showUnsyncedWarning`.

`Promise.race` will return the first promise that resolves or rejects and ignore all others where as `Promise.all` will wait until all Promises resolve or fail early if any of them reject.

Reply:
Hmm ok, I think I indeed want Promise.all then. If we ever see a version of zero we should stop and show the warning. Otherwise, keep checking.

Reply:
Updated in 5324686. Nice catch!

Reply:
good catch @beaucollins - the `race` was from my original idea where only the unsync'd notes would resolve and so we wouldn't have needed to wait for all Promises - an early abort. this was a wrong idea. :smile:

Reply:
I've learned a lot about `Promise` via this PR! :)

Reply:
@roundhill you don't need to learn that much about promises. just let your "yes" be "yes" and your "no" be "no" :wink: